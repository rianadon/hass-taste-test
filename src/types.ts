interface IgnoreAttributesForTags {
    /** tags on which to ignore the given attributes */
    tags: string[]
    /** attributes to ignore for the given tags */
    attributes: string[]
}

export interface DiffOptions {
    /** array of attributes to ignore, when given a string that attribute will be ignored on all tags
     *  when given an object of type `IgnoreAttributesForTags`, you can specify on which tags to ignore which attributes */
    ignoreAttributes: (string | IgnoreAttributesForTags)[]
    /** array of tags to ignore, these tags are stripped from the output */
    ignoreTags: string[]
    /** array of tags whose children to ignore, the children of these tags are stripped from the output */
    ignoreChildren: string[]
    /** like ignoreChildren, but applies to shadow DOM rather than direct descendants */
    ignoreShadowChildren: string[]
    /** array of attributes which should be removed when empty.
     * Be careful not to add any boolean attributes here (e.g. `hidden`) unless you know what you're doing */
    stripEmptyAttributes: string[]
    /** whether to strip attributes from Lit & Polymer with undefined values */
    stripUndefinedAttributes: boolean
}

export type DashboardOptions = Partial<{
    colorScheme: 'light' | 'dark' | 'no-preference'
    title: string
}>

export interface BrowserIntegration<E> {
    open(url: string, options: DashboardOptions): Promise<BrowserPage<E>>
    openInHeaded(url: string): Promise<void>
    close(): Promise<void>
}

export interface BrowserPage<E> {
    getNthCard(n: number): Promise<E>
    shadowHTML(element: E, options?: Partial<DiffOptions>): Promise<string>
    textContent(element: E): Promise<string>
    screenshot(element: E): Promise<Buffer>
    find(element: E, selector: string): Promise<E | null | undefined>
}

export type LovelaceResourceType = 'css' | 'js' | 'module' | 'html'

export interface LovelaceDashboardMutableParams {
    require_admin: boolean
    show_in_sidebar: boolean
    icon?: string
    title: string
}

export interface LovelaceDashboardCreateParams extends LovelaceDashboardMutableParams {
    url_path: string
    mode: 'storage'
}
