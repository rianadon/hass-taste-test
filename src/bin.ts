#!/usr/bin/env node
import { HomeAssistant } from './'
;(async () => {
    console.log('Ensuring Home Assistant packages are installed...')
    const hass = await HomeAssistant.create('recorder:', {})
    await hass.close()
    console.log('All set!')
})()
