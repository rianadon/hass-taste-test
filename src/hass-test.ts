import { spawn, ChildProcess, execFile } from 'child_process'
import { basename, join } from 'path'
import { tmpdir } from 'os'
import { mkdtemp, writeFile, rm, mkdir, copyFile, stat, symlink } from 'fs/promises'
import fetch from 'node-fetch'
import { createConnection as createTCPConnection, createServer as createTCPServer } from 'net'
import { Auth, callService, Connection, createConnection as createHAConnection, HassServiceTarget } from 'home-assistant-js-websocket'
import { createSocket } from './socket'
import { BrowserPage, DiffOptions, LovelaceDashboardCreateParams, LovelaceResourceType } from './types'
import { BrowserIntegration } from './types'
import lockfile from 'proper-lockfile'
import { promisify } from 'util'

interface HassTestOptions<E> {
    python: string
    hassArgs: string[]
    host: string
    port: number | null
    username: string
    password: string
    userLanguage: string
    userDisplayName: string
    customComponents: string[]
    integration?: BrowserIntegration<E>
}

const DEFAULT_CONFIG = (options: HassTestOptions<any>, port: number) => `
frontend:
http:
  server_host: ${options.host}
  server_port: ${port}
`

// Global port to start search from
// When multiple HassTest instances are created, they will all increment this port
// Hopefully this stops them from colliding with one another
let startPort = 8130

const exec = (command: string, args: string[]) => new Promise((resolve, reject) => {
    const proc = spawn(command, args, { stdio: 'inherit' })
    proc.on('error', err => reject(err))
    proc.on('close', code => resolve(code))
})

export default class HassTest<E> {

    private venvDir!: string
    private configDir!: string

    private chosenPort!: number
    private process!: ChildProcess
    private accessCode!: string
    private accessToken!: string
    private options: HassTestOptions<E>
    private dashboards = 0

    public ws!: Connection

    constructor(private config: string, options?: Partial<HassTestOptions<E>>) {
        this.venvDir = join(tmpdir(), 'hasstest-venv')
        this.options = {
            python: 'python3',
            hassArgs: [],
            host: '127.0.0.1',
            port: null,
            username: 'dev',
            password: 'dev',
            userLanguage: 'en',
            userDisplayName: 'Developer',
            customComponents: [],
            ...options
        }
    }

    private get path_confFile() { return join(this.configDir, 'configuration.yaml') }
    private get path_componentsDir() { return join(this.configDir, 'custom_components') }
    private get path_wwwDir() { return join(this.configDir, 'www') }
    private get path_hass() { return join(this.venvDir, 'bin/hass') }
    private get path_pip() { return join(this.venvDir, 'bin/pip') }

    /** Start the HomeAssistant server and connect to its websocket */
    public async start() {
        this.configDir = await mkdtemp(join(tmpdir(), 'hasstest-'))

        await mkdir(this.path_wwwDir)
        await mkdir(this.path_componentsDir)
        await this.linkComponents()

        const releaseLock = await this.acquireLock()
        await Promise.all([this.findPort(), this.setupVenv()])

        await writeFile(this.path_confFile , DEFAULT_CONFIG(this.options, this.chosenPort) + this.config)
        this.process = spawn(this.path_hass, ['-c', this.configDir, ...this.options.hassArgs], {
            stdio: 'inherit'
        })
        while (!(await this.isUp())) { /* wait */ }
        await releaseLock() // Release lock once the process is up, so other instances can select a different port

        await this.onboard()
    }

    /** Symlink custom component directories to the configuration directory */
    private async linkComponents() {
        await Promise.all(this.options.customComponents.map(component =>
            symlink(component, join(this.path_componentsDir, basename(component)), 'junction')
        ))
    }

    /** Wait for lockfile to become released. Forces HassTest instances to start up one at a time */
    private async acquireLock() {
        for (let iterations = 0; iterations < 50; iterations++)
            try {
                return await lockfile.lock(tmpdir(), { lockfilePath: 'hasstest.lock' })
            } catch (e) {
                if (e.code !== 'ELOCKED') throw e
                await new Promise(resolve => setTimeout(resolve, 100))
            }

        throw new Error('Timed out waiting for lockfile')
    }

     /** Find an open port to launch Home Assistant */
    private async findPort() {
        this.chosenPort = startPort++
        const server = createTCPServer()
        server.listen(this.chosenPort, this.options.host)
        const success = await new Promise((resolve, reject) => {
            server.on('error', (e: NodeJS.ErrnoException) => {
                if (e.code === 'EADDRINUSE' || e.code === 'EACCESS') resolve(false)
                else reject(e)
            })
            server.on('listening', () => {
                server.close()
                resolve(true)
            })
        })
        if (!success) await this.findPort()
    }

    /** Creates a python virtual environment and installs Home Assistant */
    private async setupVenv() {
        // Create virtual environment if it does not already exist
        if (!(await stat(this.venvDir)).isDirectory())
            await exec(this.options.python, ['-m', 'venv', this.venvDir])

        // Check if homeassistant needs an upgrade; saves a bit of time if it doesn't need one
        const latest = await fetch('https://pypi.org/pypi/homeassistant/json').then(r => r.json())
        const { stdout } = await promisify(execFile)(this.path_pip, ['freeze'])
        const installed = stdout.split('\n').find(v => v.startsWith('homeassistant=='))

        if (!installed || installed.split('==')[1] !== latest.info.version) {
            await exec(this.path_pip, ['install', '--upgrade', 'homeassistant'])
        }
    }

    /** Checks if Home Assistant is listening on its TCP port */
    private isUp = () => new Promise(resolve => {
        createTCPConnection(this.chosenPort, this.options.host).on('connect', () => {
            resolve(true)
        }).on("error", () => {
            setTimeout(() => resolve(false), 300)
        })
    })

    /** Complete onboarding and fetch short-lived access token */
    async onboard() {
        const login = await this.post('/api/onboarding/users', {
            language: this.options.userLanguage,
            name: this.options.userDisplayName,
            username: this.options.username,
            password: this.options.password
        })
        this.accessCode = login.auth_code
        await this.launchWebsocket(this.accessCode)

        // Step through the onboarding pages
        await this.post('/api/onboarding/core_config', {}, true)
        await this.post('/api/onboarding/analytics', {}, true)
        const response = await this.post('/api/onboarding/integration', {
            redirect_uri: this.url + '/?auth_callback=1'
        }, true)
        this.accessCode = response.auth_code
    }

    get url() { return `http://${this.options.host}:${this.chosenPort}` }
    get clientId() { return this.url + '/' }
    get link() { return this.customDashboard('') }

    public customDashboard(name: string, code=this.accessCode) {
        if (!code) throw new Error("You haven't logged into Home Assistant yet. Have you called hasstest.start()?")
        const state = Buffer.from(JSON.stringify({ hassUrl: this.url, clientId: this.clientId })).toString('base64')
        return this.url + '/' + name + `?auth_callback=1&code=${encodeURIComponent(code)}&state=${encodeURIComponent(state)}`
    }

    async post(url: string, body: any, authorize=false) {
        const response = await fetch(this.url + url, {
            method: 'post',
            headers: {'Content-Type': 'application/json', ...(authorize ? {'Authorization': 'Bearer ' + this.accessToken } : undefined)},
            body: JSON.stringify({ client_id: this.clientId, ...body })
        })
        return await response.json()
    }

    /** Fetch a access code from username and password */
    async fetchLoginCode() {
        const login = await this.post('/auth/login_flow', {
            handler: ["homeassistant", null],
            redirect_uri: this.url + '/?auth_callback=1'
        })
        const response = await this.post('/auth/login_flow/' + login.flow_id, {
            username: this.options.username,
            password: this.options.password
        })
        return response.result
    }

    /** Login and create tokens and websocket */
    async login() {
        this.accessCode = await this.fetchLoginCode()
        await this.launchWebsocket(this.accessCode)
    }

    private async launchWebsocket(code: string) {
        const params = new URLSearchParams()
        params.append('client_id', this.clientId)
        params.append('code', code)
        params.append('grant_type', 'authorization_code')
        const response = await fetch(this.url + '/auth/token', {
            method: 'post', body: params
        })
        const tokens = await response.json()
        this.accessToken = tokens.access_token

        const auth = new Auth({
            ...tokens,
            hassUrl: this.url,
            clientId: this.clientId,
        })

        this.ws = await createHAConnection({ auth, createSocket: async () => createSocket(auth, false), })
    }

    public async addIntegration(name: string) {
        const response = await this.post('/api/config/config_entries/flow', {
            handler: name,
            show_advanced_options: false
        }, true)
        if (!response.result) throw new Error('Multi-step integration flows are not yet suppoted')
        return response.result
    }

    public async addResource(filename: string, resourceType: LovelaceResourceType) {
        if (resourceType === 'module') {
            await copyFile(filename, join(this.path_wwwDir, basename(filename)))
        } else {
            throw new Error('Only the module type is supported for now')
        }
        await this.ws.sendMessagePromise({
            type: "lovelace/resources/create",
            res_type: resourceType,
            url: '/local/' + basename(filename)
        })
    }

    async createDashboard(options?: Partial<LovelaceDashboardCreateParams>) {
        if (!this.ws) throw new Error('Hass-test has not yet been initialized. Did you call hass.start()?')
        const id = ++this.dashboards
        const args = {
            type: "lovelace/dashboards/create",
            url_path: `lovelace-${id}`,
            mode: 'storage',
            require_admin: false,
            show_in_sidebar: true,
            title: `Dashboard ${id}`,
            ...options
        }
        await this.ws.sendMessagePromise(args)
        return args.url_path
    }

    async setDashboardView(dashboard_path: string, config: object[]) {
        if (!this.ws) throw new Error('Hass-test has not yet been initialized. Did you call hass.start()?')
        await this.ws.sendMessagePromise({
            type:"lovelace/config/save",
            url_path: dashboard_path,
            config: {
                title: "View",
                views: [ {
                    path:"default_view",
                    title:"View",
                    cards: config
                } ]
            }
        })
    }

    async callService(domain: string, service: string, serviceData?: object, target?: HassServiceTarget) {
        return await callService(this.ws, domain, service, serviceData, target)
    }

    async Dashboard(config: object[]) {
        if (!this.options.integration) throw new Error('Cannot launch a dashboard without a browser integration. Make sure to specify options.integration')
        const dashboard = await this.createDashboard()
        await this.setDashboardView(dashboard, config)
        const code = await this.fetchLoginCode()
        const page = await this.options.integration.open(this.customDashboard(dashboard, code))
        return new this.HassDashboard(this, dashboard, config, page)
    }

    /** Clean up connections and stop the HomeAssistant server */
    async close() {
        if (this.options.integration) await this.options.integration.close()
        if (this.ws) this.ws.close()
        if (this.process) {
            this.process.kill('SIGINT')
            await new Promise(resolve => this.process.on('exit', resolve))
        }
        if (this.configDir) await rm(this.configDir, { recursive: true })
    }

    public HassDashboard = class {

        public cards: HassCard<E>[] = []

        constructor (public parent: HassTest<E>, public name: string, config: object[], page: BrowserPage<E>) {
            for (let i = 0; i < config.length; i++) {
                this.cards.push(new HassCard(i, page))
            }
        }

        async link() {
            const code = await this.parent.fetchLoginCode()
            return this.parent.customDashboard(this.name, code)
        }

        async openInBrowser() {
            await this.parent.options.integration!.openInHeaded(await this.link())
        }
    }
}

export class HassCard<E> {
    private selectors: string[]

    constructor (private n: number, private page: BrowserPage<E>, selectors?: string[]) {
        this.selectors = selectors || []
    }

    private async element() {
        let el = await this.page.getNthCard(this.n)
        for (const sel of this.selectors) {
            const element = await this.page.find(el, sel)
            if (!element) throw new Error('Could not find selector ' + sel)
            el = element
        }
        return el
    }

    public narrow(selector: string) {
        return new HassCard<E>(this.n, this.page, [...this.selectors, selector])
    }

    public async html(options?: DiffOptions) {
        return await this.page.shadowHTML(await this.element(), options)
    }

    public async text() {
        return (await this.page.textContent(await this.element())).trim()
    }

    public async screenshot() {
        return await this.page.screenshot(await this.element())
    }

}

export function multiply(n: number, configFn: (i: number) => string) {
    let config = '';
    for (let i = 1; i <= n; i++) {
        config += configFn(i) + '\n';
    }
    return config
}
