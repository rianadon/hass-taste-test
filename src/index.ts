import { spawn, ChildProcess } from 'child_process'
import { basename, join } from 'path'
import { tmpdir } from 'os'
import fetch from 'node-fetch'
import * as fs from 'fs/promises'
import * as net from 'net'
import * as hass from 'home-assistant-js-websocket'
import { createSocket } from './socket'
import {
    BrowserPage,
    DashboardOptions,
    DiffOptions,
    LovelaceDashboardCreateParams,
    LovelaceResourceType,
} from './types'
import { BrowserIntegration } from './types'
import lockfile from 'proper-lockfile'
import { Collection, subscribeEntities, HassEntities } from 'home-assistant-js-websocket'
export { PlaywrightBrowser, PlaywrightElement } from './integrations/playwright'

interface HassOptions<E> {
    python: string
    hassArgs: string[]
    host: string
    port: number | null
    username: string
    password: string
    userLanguage: string
    userDisplayName: string
    customComponents: string[]
    browser?: BrowserIntegration<E>
}

interface CacheConf {
    latestHass: string
    startPort: number
}

/** Sleeps for 100 ms */
const sleep = () => new Promise((r) => setTimeout(r, 100))

const exec = (command: string, args: string[]) =>
    new Promise((resolve, reject) => {
        const proc = spawn(command, args, { stdio: 'inherit' })
        proc.on('error', (err) => reject(err))
        proc.on('close', (code) => resolve(code))
    })

export class HomeAssistant<E> {
    private venvDir!: string
    private configDir!: string

    private cache!: CacheConf
    private chosenPort!: number
    private process!: ChildProcess
    private accessCode!: string
    private accessToken!: string
    private options: HassOptions<E>
    private dashboards = 0
    private entities: HassEntities = {}

    public ws!: hass.Connection

    private constructor(options?: Partial<HassOptions<E>>) {
        this.venvDir = join(tmpdir(), 'hasstest-venv')
        this.options = {
            python: 'python3',
            hassArgs: [],
            host: '127.0.0.1',
            port: null,
            username: 'dev',
            password: 'dev',
            userLanguage: 'en',
            userDisplayName: 'Developer',
            customComponents: [],
            ...options,
        }
    }

    private path_confFile = () => join(this.configDir, 'configuration.yaml')
    private path_componentsDir = () => join(this.configDir, 'custom_components')
    private path_wwwDir = () => join(this.configDir, 'www')
    private path_hass = () => join(this.venvDir, 'bin/hass')
    private path_pip = () => join(this.venvDir, 'bin/pip')
    private path_cache = () => join(tmpdir(), 'hasstest-cache')

    /** Create and connect to an isolated Home Assistant instance */
    public static async create<E>(config: string, options?: Partial<HassOptions<E>>) {
        const hass = new HomeAssistant<E>(options)
        await hass.start(config)
        return hass
    }

    /** Log into an existing Home Assistant instance */
    public static async connect<E>(options?: Partial<HassOptions<E>>) {
        const hass = new HomeAssistant<E>(options)
        await hass.login()
        return hass
    }

    /** Start the Home Assistant server and connect to its websocket */
    private async start(config: string) {
        this.configDir = await fs.mkdtemp(join(tmpdir(), 'hasstest-'))

        // Step 1: Create necessary directories
        await Promise.all([
            fs.mkdir(this.path_wwwDir()),
            fs.mkdir(this.path_componentsDir()).then(() => this.linkComponents()),
        ])

        // Step 2: Create venv and sort out networking under lock
        const releaseLock = await this.acquireLock()
        this.cache = await this.readCache()
        await Promise.all([
            this.findPort().then(() => this.writeYAMLConfiguration(config)),
            this.setupVenv(),
        ])
        await fs.writeFile(this.path_cache(), JSON.stringify(this.cache))
        await releaseLock()

        // Step 3: Start Home Assistant and onboard
        this.process = spawn(
            this.path_hass(),
            ['-c', this.configDir, '-v', ...this.options.hassArgs],
            {
                stdio: 'inherit',
            },
        )
        this.process.on('error', (e) => {
            console.error(
                `----\nHome Assistant is having trouble starting. This is likely due to a problem with your virtual environment; try removing the venv folder at ${this.venvDir} and re-running hass-taste-test\n---`,
            )
            throw e
        })
        await new Promise<void>(async (resolve, reject) => {
            // Wait for Home Assistant to start, but quit early if the process exits
            this.process.on('close', (code) => {
                if (code != 0) reject(new Error(`Home Assistant exited with code ${code}`))
            })
            while (!(await this.isUp())) await sleep()
            resolve()
        })
        await this.onboard()
    }

    /** Write configuration.yaml */
    private async writeYAMLConfiguration(additionalConfig: string) {
        const config = [
            'frontend:',
            'http:',
            `  server_host: ${this.options.host}`,
            `  server_port: ${this.chosenPort}`,
            additionalConfig,
        ].join('\n')
        fs.writeFile(this.path_confFile(), config)
    }

    /** Symlink custom component directories to the configuration directory */
    private async linkComponents() {
        await Promise.all(
            this.options.customComponents.map((component) =>
                fs.symlink(
                    component,
                    join(this.path_componentsDir(), basename(component)),
                    'junction',
                ),
            ),
        )
    }

    /** Wait for lockfile to become released. Forces HassTest instances to start up one at a time */
    private async acquireLock() {
        for (let iterations = 0; iterations < 600; iterations++)
            try {
                return await lockfile.lock(tmpdir(), { lockfilePath: 'hasstest.lock' })
            } catch (e: any) {
                if (e.code !== 'ELOCKED') throw e
                await sleep()
            }
        throw new Error('Timed out waiting for lockfile')
    }

    /** Find an open port to launch Home Assistant */
    private async findPort() {
        this.chosenPort = this.cache.startPort++
        const server = net.createServer()
        server.listen(this.chosenPort, this.options.host)
        const success = await new Promise((resolve, reject) => {
            server.on('error', (e: NodeJS.ErrnoException) => {
                if (e.code === 'EADDRINUSE' || e.code === 'EACCESS') resolve(false)
                else reject(e)
            })
            server.on('listening', () => {
                server.close()
                resolve(true)
            })
        })
        if (!success) await this.findPort()
    }

    /** Creates a python virtual environment and installs Home Assistant */
    private async setupVenv() {
        // Create virtual environment if it does not already exist
        await fs
            .access(this.venvDir)
            .catch(() => exec(this.options.python, ['-m', 'venv', this.venvDir]))

        // Check if homeassistant needs an upgrade; saves a bit of time if it doesn't need one
        const installed = await this.hassVersion()

        if (!installed || installed !== this.cache.latestHass) {
            await exec(this.path_pip(), ['install', '--upgrade', 'wheel'])
            await exec(this.path_pip(), ['install', '--upgrade', 'mutagen', 'homeassistant'])
        }
    }

    private async readCache(): Promise<CacheConf> {
        try {
            const versionFile = await fs.stat(this.path_cache())
            if (Date.now() - versionFile.mtimeMs < 2000) {
                return await JSON.parse(await fs.readFile(this.path_cache(), 'utf-8'))
            }
        } catch (e) {}
        return {
            latestHass: await this.latestHAVersion(),
            startPort: 8130,
        }
    }

    /** Fetch the latest hass version from pypi */
    private async latestHAVersion(): Promise<string> {
        const latest = await fetch('https://pypi.org/pypi/homeassistant/json').then(
            (r) => r.json() as any,
        )
        return latest.info.version
    }

    /** Finds the version of Home Assistant installed, if any */
    private async hassVersion(): Promise<string | null> {
        const libFolders = await fs.readdir(join(this.venvDir, 'lib'))
        const libs = await Promise.all(
            libFolders.map((f) => fs.readdir(join(this.venvDir, 'lib', f, 'site-packages'))),
        )
        const homeAssistant = libs
            .flat()
            .map((f) => f.match(/^homeassistant-(.*)\.dist-info/))
            .find((f) => f !== null)
        return homeAssistant ? homeAssistant[1] : null
    }

    /** Checks if Home Assistant is listening on its TCP port */
    private isUp = () =>
        new Promise((resolve) => {
            net.createConnection(this.chosenPort, this.options.host)
                .on('connect', () => resolve(true))
                .on('error', () => resolve(false))
        })

    /** Complete onboarding and fetch short-lived access token */
    private async onboard() {
        let login
        for (let tries = 0; tries < 20; tries++) {
            try {
                login = await this.post('/api/onboarding/users', {
                    language: this.options.userLanguage,
                    name: this.options.userDisplayName,
                    username: this.options.username,
                    password: this.options.password,
                })
                if (login.auth_code) break
                else await sleep()
            } catch (e) {
                console.warn(e)
                await sleep()
            }
        }
        if (!login.auth_code) throw new Error('No auth code found')
        this.accessCode = login.auth_code
        await this.launchWebsocket(this.accessCode)

        // Step through the onboarding pages
        await this.post('/api/onboarding/core_config', {}, true)
        await this.post('/api/onboarding/analytics', {}, true)
        const response = await this.post(
            '/api/onboarding/integration',
            { redirect_uri: this.url + '/?auth_callback=1' },
            true,
        )
        this.accessCode = response.auth_code
    }

    private get url() {
        return `http://${this.options.host}:${this.chosenPort}`
    }
    private get clientId() {
        return this.url + '/'
    }
    public get link() {
        return this.customDashboard('')
    }

    public customDashboard(path: string, code = this.accessCode) {
        if (!code)
            throw new Error(
                'You have not logged into Home Assistant yet. Have you called hasstest.start()?',
            )
        const state = Buffer.from(
            JSON.stringify({ hassUrl: this.url, clientId: this.clientId }),
        ).toString('base64')
        return `${this.url}/${path}?auth_callback=1&code=${encodeURIComponent(
            code,
        )}&state=${encodeURIComponent(state)}`
    }

    public async post(url: string, body: any, authorize = false) {
        const response = await fetch(this.url + url, {
            method: 'post',
            headers: {
                'Content-Type': 'application/json',
                ...(authorize ? { Authorization: 'Bearer ' + this.accessToken } : undefined),
            },
            body: JSON.stringify({ client_id: this.clientId, ...body }),
        })
        return (await response.json()) as any
    }

    /** Fetch a access code from username and password */
    private async fetchLoginCode() {
        const login = await this.post('/auth/login_flow', {
            handler: ['homeassistant', null],
            redirect_uri: this.url + '/?auth_callback=1',
        })
        const response = await this.post('/auth/login_flow/' + login.flow_id, {
            username: this.options.username,
            password: this.options.password,
        })
        if (response.error) throw new Error(`Login code returned with error ${response.error}`)
        return response.result
    }

    /** Login and create tokens and websocket */
    public async login() {
        this.accessCode = await this.fetchLoginCode()
        await this.launchWebsocket(this.accessCode)
    }

    private async launchWebsocket(code: string) {
        const params = new URLSearchParams()
        params.append('client_id', this.clientId)
        params.append('code', code)
        params.append('grant_type', 'authorization_code')
        const response = await fetch(this.url + '/auth/token', {
            method: 'post',
            body: params,
        })
        const tokens = (await response.json()) as any
        if (tokens.error) throw new Error(`Tokens returned with error ${tokens.error}`)
        this.accessToken = tokens.access_token

        const auth = new hass.Auth({
            ...tokens,
            hassUrl: this.url,
            clientId: this.clientId,
        })

        this.ws = await hass.createConnection({
            auth,
            createSocket: async () => createSocket(auth, false),
        })
        subscribeEntities(this.ws, (e) => (this.entities = { ...this.entities, ...e }))
    }

    public async addIntegration(name: string) {
        const response = await this.post(
            '/api/config/config_entries/flow',
            {
                handler: name,
                show_advanced_options: false,
            },
            true,
        )
        if (response.message)
            throw new Error(
                `Error configuring ${name} integration. Check the logs: ${response.message}`,
            )
        if (!response.result) throw new Error('Multi-step integration flows are not yet suppoted')
        return response.result
    }

    public async addResource(filename: string, resourceType: LovelaceResourceType) {
        if (resourceType === 'module') {
            await fs.copyFile(filename, join(this.path_wwwDir(), basename(filename)))
        } else {
            throw new Error('Only the module type is supported for now')
        }
        await this.ws.sendMessagePromise({
            type: 'lovelace/resources/create',
            res_type: resourceType,
            url: '/local/' + basename(filename),
        })
    }

    public async createDashboard(options?: Partial<LovelaceDashboardCreateParams>) {
        if (!this.ws)
            throw new Error('Hass-test has not yet been initialized. Did you call hass.start()?')
        const id = ++this.dashboards
        const args = {
            type: 'lovelace/dashboards/create',
            url_path: `lovelace-${id}`,
            mode: 'storage',
            require_admin: false,
            show_in_sidebar: true,
            ...options,
            title: options?.title || `Dashboard ${id}`,
        }
        await this.ws.sendMessagePromise(args)
        return args.url_path
    }

    async setDashboardView(dashboard_path: string, config: object[]) {
        if (!this.ws)
            throw new Error('Hass-test has not yet been initialized. Did you call hass.start()?')
        await this.ws.sendMessagePromise({
            type: 'lovelace/config/save',
            url_path: dashboard_path,
            config: {
                title: 'View',
                views: [{ path: 'default_view', title: 'View', cards: config }],
            },
        })
    }

    async callService(
        domain: string,
        service: string,
        serviceData?: object,
        target?: hass.HassServiceTarget,
    ) {
        return await hass.callService(this.ws, domain, service, serviceData, target)
    }

    async Dashboard(config: object[], options?: DashboardOptions) {
        if (!this.options.browser)
            throw new Error(
                'Cannot launch a dashboard without a browser integration. Make sure to specify options.integration',
            )
        const dashboard = await this.createDashboard({ title: options?.title })
        await this.setDashboardView(dashboard, config)
        const code = await this.fetchLoginCode()
        const page = await this.options.browser.open(
            this.customDashboard(dashboard, code),
            options || {},
        )
        return new this.HassDashboard(this, dashboard, config, page)
    }

    async states() {
        return this.entities
    }

    /** Clean up connections and stop the HomeAssistant server */
    async close() {
        if (this.options.browser) await this.options.browser.close()
        if (this.ws) this.ws.close()
        if (this.process) {
            this.process.kill('SIGINT')
            await new Promise((resolve) => this.process.on('exit', resolve))
        }
        if (this.configDir) await fs.rm(this.configDir, { recursive: true })
    }

    public HassDashboard = class {
        public cards: HassCard<E>[] = []

        constructor(
            public parent: HomeAssistant<E>,
            public path: string,
            config: object[],
            public page: BrowserPage<E>,
        ) {
            for (let i = 0; i < config.length; i++) {
                this.cards.push(new HassCard(i, page))
            }
        }

        async link() {
            const code = await this.parent.fetchLoginCode()
            return this.parent.customDashboard(this.path, code)
        }

        async openInBrowser() {
            await this.parent.options.browser!.openInHeaded(await this.link())
        }
    }
}

export class HassCard<E> {
    private selectors: string[]

    constructor(
        private n: number,
        private page: BrowserPage<E>,
        selectors?: string[],
    ) {
        this.selectors = selectors || []
    }

    public async element() {
        let el = await this.page.getNthCard(this.n)
        for (const sel of this.selectors) {
            const element = await this.page.find(el, sel)
            if (!element) throw new Error('Could not find selector ' + sel)
            el = element
        }
        return el
    }

    public narrow(selector: string) {
        return new HassCard<E>(this.n, this.page, [...this.selectors, selector])
    }

    public async html(options?: Partial<DiffOptions>) {
        return await this.page.shadowHTML(await this.element(), options)
    }

    public async text() {
        return (await this.page.textContent(await this.element())).trim()
    }

    public async screenshot() {
        return await this.page.screenshot(await this.element())
    }
}

export function multiply(n: number, configFn: (i: number) => string) {
    let config = ''
    for (let i = 1; i <= n; i++) {
        config += configFn(i) + '\n'
    }
    return config
}
