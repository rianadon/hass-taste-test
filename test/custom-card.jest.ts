import { HomeAssistant, PlaywrightIntegration, PlaywrightElement } from '../src'
import { toMatchImageSnapshot } from 'jest-image-snapshot'

expect.extend({ toMatchImageSnapshot })

const CONFIGURATION_YAML = `
input_boolean:
  test:
`

let hass: HomeAssistant<PlaywrightElement>

beforeAll(async () => {
    hass = new HomeAssistant(CONFIGURATION_YAML, {
        integration: new PlaywrightIntegration(process.env.BROWSER || 'firefox'),
    })
    await hass.start()
    await hass.addResource(__dirname + '/resources/custom-card.js', 'module')
}, 30000)
afterAll(async () => await hass.close())

it('Custom Card', async () => {
    const dashboard = await hass.Dashboard([
        { type: 'custom:content-card-example', entity: 'input_boolean.test' },
    ])
    const card = dashboard.cards[0]
    expect(await card.html()).toMatchSnapshot()
    expect(await card.screenshot()).toMatchImageSnapshot()
})
