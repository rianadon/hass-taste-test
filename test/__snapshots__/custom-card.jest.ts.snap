// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Card 1`] = `
"<content-card-example>
  <ha-card header="Example-card" raised="false">
    <h1 class="card-header">
      Example-card
    </h1>
    <slot></slot>
    <div class="card-content">
      The state of input_boolean.test is off!
      <br>
      <br>
      <img src="data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 350 150'%3E%3Crect fill='%23cccccc' width='350' height='150'%3E%3C/rect%3E%3Ctext x='50%25' y='50%25' dominant-baseline='middle' text-anchor='middle' font-family='monospace' font-size='26px' fill='%23333333'%3E350x150%3C/text%3E%3C/svg%3E">
    </div>
  </ha-card>
</content-card-example>
"
`;
